runtime/sema/check_casting_expression.go

```diff
@@ -412,3 +421,204 @@ func FailableCastCanSucceed(subType, superType Type) bool {
 
 	return true
 }
+
+// IsCastRedundant checks whether a simple cast is redundant.
+// Checks for two cases:
+//    - Case I: Contextually expected type is same as casted type (target type).
+//    - Case II: Expression is self typed, and is same as the casted type (target type).
+func IsCastRedundant(expr ast.Expression, exprInferredType, targetType, expectedType Type) bool {
+	if expectedType != nil && expectedType.Equal(targetType) {
+		return true
+	}
+
+	checkCastVisitor := &CheckCastVisitor{}
+
+	return checkCastVisitor.isCastRedundant(expr, exprInferredType, targetType)
```

Comment:
Great idea to factor the casting check into its own type, though I think there is no need for a pointer type here (only the mutating functions need a pointer receiver):

```suggestion
	return CheckCastVisitor{}.isCastRedundant(expr, exprInferredType, targetType)
```

Reply:
`CheckCastVisitor` maintains an updates the `exprInferredType` and `targetType` as it goes into nested expressions (e.g: array/dictionary literals). So going to need a pointer receiver.

Reply:
Right, the mutating functions should have a pointer receiver, but the value here itself doesn't have to be, see e.g. https://go.dev/play/p/iF22kMUH_MV

Reply:
`(d *CheckCastVisitor) IsRedundantCast()` is actually the mutating function, and is recursively invoked by the others.